src/Interval.h

```diff
@@ -0,0 +1,301 @@
+/*
+ * LSST Data Management System
+ * Copyright 2014-2015 AURA/LSST.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <https://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#ifndef LSST_SG_INTERVAL_H_
+#define LSST_SG_INTERVAL_H_
+
+/// \file
+/// \author Serge Monkewitz
+/// \brief This file defines a template representing closed real intervals.
+
+#include <algorithm>
+#include <ostream>
+
+#include "SpatialRelation.h"
+
+
+namespace lsst {
+namespace sg {
+
+/// `Interval` represents a closed interval of the real numbers
+/// by its upper and lower bounds. It is parameterized by the Scalar
+/// type, which must be constructible from a double.
+///
+/// An interval with identical upper and lower bounds contains
+/// a single point, and is equal to that scalar bound.
+///
+/// An interval with an upper bound less than its lower bound is
+/// empty (contains no points), as are intervals with NaN as
+/// either bound.
+template <typename Derived, typename Scalar>
+class Interval {
+public:
+    /// This constructor creates an empty interval.
+    Interval() : _a(1.0), _b(0.0) {}
+    /// This constructor creates a closed interval containing only x.
+    explicit Interval(Scalar x) : _a(x), _b(x) {}
+    /// This constructor creates an interval from the given endpoints.
+    Interval(Scalar x, Scalar y) : _a(x), _b(y) {}
+
+    /// Two closed intervals are equal if their endpoints are the same, or
+    /// both are empty.
+    bool operator==(Interval const & i) const {
+        return (_a == i._a && _b == i._b) || (i.isEmpty() && isEmpty());
+    }
+    bool operator!=(Interval const & i) const { return !(*this == i); }
+
+    /// A closed interval is equal to a point x if both endpoints equal x.
+    bool operator==(Scalar x) const {
+        return (_a == x && _b == x) || (x != x && isEmpty());
+    }
+    bool operator!=(Scalar x) const { return !(*this == x); }
+
+    ///@{
+    /// `a` returns the lower endpoint of this interval. The return value
+    /// for empty intervals is arbitrary.
+    Scalar a() const { return _a; }
+    Scalar & a() { return _a; }
+    ///@}
```

Comment:
I think the value of the in-place shrink/expand outweighs any problems that might appear in Python later, which I think can be at least partially worked around by just not exposing those methods to Python.

I should be a bit less cryptic on this point, though: I think composition in Python just works much better with immutable objects (this is much of why Python has immutable strings and tuples, I think), because it also tends to share everything.  That's not always a good idea in C++ of course, but you can also get some of the benefits by wrapping a mutable C++ class into something that's immutable from the Python side.  For instance, that lets C++ code return a contained object by const reference without worrying that a Python user could directly modify it.